adapter/adapter_roundrobin.c

```diff
@@ -0,0 +1,102 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/*
+ * Copyright 2023-2024 Huawei Technologies Co.,Ltd. All rights reserved.
+ * Copyright 2023-2024 Linaro ltd.
+ */
+
+#include "adapter_private.h"
+
+struct rr_adapter_ctx {
+	uint32_t send_idx;
+	uint32_t recv_idx;
+};
+
+static int uadk_adapter_rr_init(struct wd_alg_driver *adapter)
+{
+	struct uadk_adapter_ctx *ctx = (struct uadk_adapter_ctx *)adapter->priv;
+	struct rr_adapter_ctx *priv = (struct rr_adapter_ctx *)ctx->priv;
+
+	if (!priv) {
+		priv = calloc(1, sizeof(*priv));
+		if (!priv)
+			return -ENOMEM;
+
+		ctx->priv = priv;
+	}
+
+	return 0;
+}
+
+static void uadk_adapter_rr_exit(struct wd_alg_driver *adapter)
+{
+	struct uadk_adapter_ctx *ctx = (struct uadk_adapter_ctx *)adapter->priv;
+
+	if (ctx->priv) {
+		free(ctx->priv);
+		ctx->priv = NULL;
+	}
+}
+
+/* fixme, how to ensure send and recv are matched in async mode */
```

Comment:
There are several issues that need to be analyzed:
1. Adapter scheduling in asynchronous mode cannot currently match.
2. The current adapter’s driver scheduling can only implement the RR scheduling algorithm.
3. In a UADK framework, there are two levels of scheduling (ctx
scheduler and driver adapter) are unreasonable and should be fused

Reply:
> There are several issues that need to be analyzed:
> 
> 1. Adapter scheduling in asynchronous mode cannot currently match.
Do you mean one transaction is split into several times, some use opsA, some use opsB?

> 2. The current adapter’s driver scheduling can only implement the RR scheduling algorithm.
Have realized threshold policy, but need to get packet size, via changing the struct.

> 3. In a UADK framework, there are two levels of scheduling (ctx
>    scheduler and driver adapter) are unreasonable and should be fused
They are different thing, adapter only choose ops, ie, choose driver.
The ctx is hardware resource, I think it likes memory, alloc and free, no need schedule.
Anyway, it's better to split.

Reply:
> > There are several issues that need to be analyzed:
> > 
> > 1. Adapter scheduling in asynchronous mode cannot currently match.
> >    Do you mean one transaction is split into several times, some use opsA, some use opsB?
>  

No. What I mean is that currently only synchronous mode is implemented in the adapter code.
In asynchronous mode, if multiple drivers are supported at the same time. It may not be possible to recycle the message.
> > 2. The current adapter’s driver scheduling can only implement the RR scheduling algorithm.
> >    Have realized threshold policy, but need to get packet size, via changing the struct.
> 

If the adapter needs to actually implement software and hardware multiple drivers to send and receive messages in a mixed manner. It needs to support the device corresponding to the driver to obtain the device load in real time. Send messages based on load data
> > 3. In a UADK framework, there are two levels of scheduling (ctx
> >    scheduler and driver adapter) are unreasonable and should be fused
> >    They are different thing, adapter only choose ops, ie, choose driver.
> >    The ctx is hardware resource, I think it likes memory, alloc and free, no need schedule.
> >    Anyway, it's better to split.
My point of view is exactly the opposite of yours. ctx is allocated from the device, and the driver is also matched one-to-one with the device. Therefore, after ctx is determined, the corresponding device is also determined. At the same time, the driver must also be determined. Therefore, when the scheduler schedules ctx, it actually completes the driver selection. Therefore, the functions of the adapter can be merged into the scheduler.

Reply:
> > > There are several issues that need to be analyzed:
> > > 
> > > 1. Adapter scheduling in asynchronous mode cannot currently match.
> > >    Do you mean one transaction is split into several times, some use opsA, some use opsB?
> 
> No. What I mean is that currently only synchronous mode is implemented in the adapter code. In asynchronous mode, if multiple drivers are supported at the same time. It may not be possible to recycle the message.
> 
> > > 2. The current adapter’s driver scheduling can only implement the RR scheduling algorithm.
> > >    Have realized threshold policy, but need to get packet size, via changing the struct.
> 
> If the adapter needs to actually implement software and hardware multiple drivers to send and receive messages in a mixed manner. It needs to support the device corresponding to the driver to obtain the device load in real time. Send messages based on load data
> 
> > > 3. In a UADK framework, there are two levels of scheduling (ctx
> > >    scheduler and driver adapter) are unreasonable and should be fused
> > >    They are different thing, adapter only choose ops, ie, choose driver.
> > >    The ctx is hardware resource, I think it likes memory, alloc and free, no need schedule.
> > >    Anyway, it's better to split.
> 
> My point of view is exactly the opposite of yours. ctx is allocated from the device, and the driver is also matched one-to-one with the device. Therefore, after ctx is determined, the corresponding device is also determined. At the same time, the driver must also be determined. Therefore, when the scheduler schedules ctx, it actually completes the driver selection. Therefore, the functions of the adapter can be merged into the scheduler.

For example, the data is coming, big packet goto hardware, small packet goto sw/cpu instruction.
But we don't know the data is big or small in the beginning, so can not decide at first.

Let's discuss in the meetting with Chinese to be more clear.